Add with score functions to MMR for chroma

Hexecu opened the pull request:
## 📝 **Description**

This PR introduces two new functions to the Chroma vectorstore:

    max_marginal_relevance_search_by_vector_with_scores
    max_marginal_relevance_search_with_scores

These functions extend the existing Maximal Marginal Relevance (MMR) functionality. In addition to returning the most relevant documents, they also return their associated similarity scores. This added context enables further customization and is particularly useful for applications requiring more advanced sorting or filtering algorithms.



## 🎯 **Usecase**

The extended functionality will be especially beneficial for applications that require advanced sorting or filtering mechanisms. My specific use-case involves using a Reciprocal Rank Fusion (RRF) system across multiple queries on the same topic. The inclusion of similarity scores alongside documents facilitates this kind of advanced application.


## 🛠 **Dependencies**

None. This feature does not introduce any new dependencies.



## 👥 **Maintainers**

@rlancemartin, @eyurtsev, I would highly appreciate your feedback on these additions.

Hexecu committed:
Add with score functions to MMR for chroma

baskaryan commented on libs/langchain/langchain/vectorstores/chroma.py:
```diff
@@ -474,6 +474,93 @@ def max_marginal_relevance_search(
         )
         return docs
 
+    def max_marginal_relevance_search_by_vector_with_scores(
```
can we update max_marginal_relevance_search_by_vector to call this

Hexecu replied:
Thank you for the suggestion! Instead of having max_marginal_relevance_search_by_vector call max_marginal_relevance_search_by_vector_with_scores, which would have added an avoidable loop to strip the scores from the results, I chose to create a shared function to reduce redundant code. This function directly handles the candidates prior to iteration.

Hexecu replied:
@baskaryan  I've added the docstring to the new shared function and reformatted the code. Please let me know if there's anything else I can improve. I appreciate the time you've taken to review this!

Hexecu replied:
EDIT: I revisited my earlier implementation and realized it wasn't optimal. I've made the necessary adjustments and everything should be in order now. Thanks for your patience.

baskaryan reviewed (commented):
cc @jeffchuber

Hexecu committed:
Update MMR with scores functions chroma

Hexecu committed:
refactor: Use underscore for unused variables in return

Adjusted the variable name from 'score' to '_' to maintain code consistency throughout the file.

Hexecu committed:
Merge branch 'master' into d.leopardi/add_mmr_with_score_chroma

Hexecu commented:
Is there anything more I can do to improve this PR or is it good to go?

Hexecu commented:
@baskaryan  Hi, would you kindly review this pull request when you have a moment? 

Thank you!
